src/kbmod/standardizers/butler_standardizer.py

```diff
@@ -0,0 +1,265 @@
+import astropy.nddata as bitmask
+from astropy.wcs import WCS
+
+import numpy as np
+
+from scipy.signal import convolve2d
+
+from kbmod.standardizer import Standardizer
+from kbmod.search import layered_image, raw_image, psf
+
+
+__all__ = ["ButlerStandardizer"]
+
+
+# ButlerStd has been inherited from the Standardizer even though at the moment
+# it shares (duplicates) a lot of functionality already in the FitsStandardizer
+# I suspect we will be replacing much of this functionality later on with
+# quicker Butler queries instead of reading all of the exposures as a ExposureF
+# objects (which is slow) at which point the two classes should diverge.
+class ButlerStandardizer(Standardizer):
+
+    name="ButlerStandardizer"
+
+    mask_flags = ["BAD", "CLIPPED", "CR", "CROSSTALK", "EDGE", "NO_DATA", "SAT",
+                  "SENSOR_EDGE", "SUSPECT"]
+    """List of flags that will be masked. The flag map is extracted from the
+    Exposure object itself."""
+
+    @classmethod
+    def canStandardize(self, tgt):
+        # this is pretty hacky - but I'm not importing the entire stack to get
+        # a simple isinstance comparison
+        if "datasetref" in str(type(tgt)).lower():
+            return True, tgt
+        return False, []
+
+    def __init__(self, butler, datasetRefs=None, id=None, **kwargs):
+        # this requires an invovled logic sequence where
```

Comment:
Should we be more prescriptive throughout? A lot of the complexity here results from accepting a wide range of inputs. Is it worth considering less flexibility for more simplicity?

Reply:
I'm not sure what you mean exactly. There are a few moving parts here. The instantiation, the serialization and then future-proofing. 

The current implementation covers cases 1, 4.1 and 4.2 and spans lines 70-77 lines. So I don't know if I'd call that complicated. 

The comment is super long and mostly is a note to myself to pay attention how to map butler interface to our abstraction. The butler is complicated interface, most of which we don't support and I don't know which bits we should. Like if we have to find a way to support case 2) - I hadn't written it out but there's probably half a dozen different ways to query a butler down to `DatasetRefs`, or `DatasetIDs` and most of them can be combined. So I don't know what cases to support, or how to abstract the interface in a way that might not require us to rewrite it once we get some users to actually use it. This leaves us a bit exposed because we can't future-proof this implementation in the same way that  `FitsStandardizer` does (pretty much covers all possible cases).

Maybe you're asking why is `DataID` even here? That's the third problem: we're not always in control of the serialization of objects from the Rubin's Stack.                
We want to serialize `ImageCollection` for share-ability and "transport" (remember the ping-pong match between Hyak and Epyc for processing of large datasets). But the serialization happens in `ImageCollection`, which itself can't know about the individual `Standardizers`. Like, what happens in the cases when we can't construct them because data doesn't exist on that machine or the internet isn't available for a remote Butler - we couldn't make the Standardizers to get them to serialize themselves!                
Since we can `get_standardizer` of a row in `ImageCollection`, we need to serialize sufficient information to construct the Standardizer back out at the other end again.              
So we're in a situation where `ImageCollection` has to know how to serialize itself, that can't be interfaced by the `Standardizers`, and we can't can't guarantee they produce only serializable values, or that they standardize sufficient data to instantiate themselves. So in this case, to demonstrate that, I serialize "DataIds" in the default `standardizeMetadata` method and then resolve the `DataId`s to `DatasetRef`s on the round-trip. Is this a problem? I don't know. It certainly feels like an edge-case.            
Why just an edge-case? Because we provide the default standardizers for most use-cases, we can always guarantee that we will always standardize at least sufficient information to instantiate them by default; therefore for all extensions of them also. We might enforce that in tests, but if a user inherits from the most general `Standardizer` and attempts to implement things from scratch and skips the tests - they just might shoot themselves in the foot when they try to write it to a file. So this is like feature not a bug, in the sense "we take care of you so you don't have to"?          

The last way I think in which I can interpret your questions is that you generally find the `__init__`s not prescriptive enough? I don't know, I feel like this one is ok, `Standardizer` really just wants the `location`, `FitsStandardizer` and `SingleExtensionFits` also. `MultiExtensionFits` has a new magic kwarg - `set_exts`, which is sort of a micro-optimization for a default functionality because it's expensive and I don't want to have to run it.            
`ButlerStandardizer` takes `DatasetRef`s and `DataID`s but this is a consequence of the fact that we support retrieving `get_standardizer` from `ImageCollection`. Maybe that needs to go, but then how does `toImageStack` method work?                     
There is one complex init - the `ImageCollection`. I've seen worse, but that's not really an excuse. Unfortunately I'm not sure how avoidable that is. The issue there is that we serialize internals into a data-table and then have to unserialize them and set them as some class properties. This requires to check if all the columns exist, and then get the column values and properly construct these internal objects. Thankfully there's only a very small limited set of them that exist, but it does produce a cacophony of `if`s I'm not a fan of either. I'm happy to hear suggestions, since off the top of my head I wouldn't know how to reduce that. I can refactor it into multiple sub-inits for separation and readability?

Reply:
I was actually talking about the first part. I saw the detail in the comment and thought some of that was placeholder for code that needed to be added. If lines 70-77 already implement all of this, then I have no concerns.